structuredheadings/plugin.js

```diff
@@ -730,31 +729,33 @@
 
           return [CKEDITOR.plugins.structuredheadings.getCurrentBlockFromPath(editor)];
         },
-
-        exec: function (editor, value) {
-          var selection = editor.getSelection();
-          var headings = CKEDITOR.plugins.structuredheadings.getHeadingsInSelection(
-                editor,
-                selection
-              );
+        exec: function (editor, value) { // eslint-disable-line max-statements
+          var headingIterator = CKEDITOR.plugins.structuredheadings.getHeadingIteratorForSelection(
+            editor
+          );
           var func;
+          var heading = headingIterator();
 
           if (value === "clear") {
             func = this.clearAutonumberClassesForHeading.bind(this, editor);
           } else {
             func = this.setAutonumberClassesForHeading.bind(this, editor, value);
 
              // prep the html for the collapsed, not in a heading case
-            if (headings === null) {
-              // put the new heading into the headings array
-              headings = this.handleCollapsedSelection(editor);
+            if (heading === null) {
+              // handle the collapsed selection by matching and converting to heading
+              this.handleCollapsedSelection(editor);
+              // then get a new iterator
+              headingIterator = CKEDITOR.plugins.structuredheadings.getHeadingIteratorForSelection(
+                editor
+              );
+              heading = headingIterator();
             }
           }
 
-          if (headings) {
-            for (var i = 0; i < headings.length; i++) {
-              func(headings[i]);
-            }
+          while (heading) {
+            func(heading);
```

Comment:
Maybe `clearOrNumber` is a better name than `func`? This looked really fishy based on the current name.